### ISSUE TITLE
Better deduction of auto

### ISSUE THREAD

**Issue by zalox** [cppit/jucipp#174]
I have this use case:

```c++
class foo {
 foo* get_ptr();
};

// main
auto ptr = foo.get_ptr();
```
If I hover ptr now, the type is auto, but i'd like it to be foo*. I realise that ptr is of type auto, but it would be more convenient in my opinion to force it to be the return value type of the expression. Any thoughts?

**Comment by milleniumbug** [comment]
+1. You can get the type if you hover on any usage of `ptr`, I assume the reason it doesn't work here is because at this point of parsing the type is unknown. If one were to query the type of `ptr` after the declaration...

**Comment by zalox** [comment]
The solution would be to use the semantics from `foo.get_ptr();` somehow. My only concern is that it might reduce snappyness from the tooltip.

**Comment by eidheim** [comment]
The problem lies in libclang. `clang_getCanonicalType(clang_getCursorType(cursor));` where `cursor` corresponds to `ptr` in your example still results in `auto`. I tried some hours last year to resolve this without any luck, but there might still be a solution. Also, it might be this problem is solved in a later libclang version. 

Anyway, you get the type when hovering/moving cursor to `get_ptr` in the last line. It might also be that it is possible to create a custom workaround for this, but these workarounds are always risky. Best thing might be to rely on libclang and maybe find a way to do this, if it is not fixed already in the latest libclang version.

**Comment by eidheim** [comment]
It seems like llvm has received a commit to solve this: https://llvm.org/bugs/show_bug.cgi?id=18669.  Hopefully this is fixed in a future version of libclang (if not fixed already in the latest version).

**Comment by eidheim** [comment]
Sorry, I really must learn to read bugzilla, https://llvm.org/bugs/show_bug.cgi?id=18669 is resolved. However, I'll see if I can improve the type information for templated types (which can be really ugly). Putting this in my TODO.

**Comment by eidheim** [comment]
Confirmed that this issue is fixed in llvm 3.8.
